### PR TITLE
Adds endpoints for showing configured spinnaker deployments

### PR DESCRIPTION
**Pull request by lwander**
@ttomsu or @jtk54 PTAL

### CODE REVIEW

**Comment by lwander** [comment]
deployment

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/providers/kubernetes/KubernetesAccount.java [hunk 1]
```diff
@@ -39,4 +39,12 @@
   List<String> namespaces = new ArrayList<>();
   @ValidateField(validators = {ValidateNotNull.class})
   List<DockerRegistryReference> dockerRegistries = new ArrayList<>();
+
+  public String getKubeconfigFile() {
+    if (kubeconfigFile == null || kubeconfigFile.isEmpty()) {
+      return System.getProperty("user.home") + "/.kube/config";
+    } else {
```
Comment (ttomsu) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/providers/kubernetes/KubernetesAccount.java]: No need for `else`

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ConfigController.java [hunk 2]
```diff
@@ -37,4 +37,14 @@
   Halconfig config() throws UnrecognizedPropertyException {
     return halyardConfig.getConfig();
   }
+
+  @RequestMapping(value = "/currentDeployment", method = RequestMethod.GET)
+  String currentDeployment() throws UnrecognizedPropertyException {
+    Halconfig halconfig = halyardConfig.getConfig();
```
Comment (ttomsu) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ConfigController.java]:
```
Halconfig halconfig = halyardConfig.getConfig();
return halconfig == null ? null : halconfig.getCurrentDeployment();
```

Makes you miss groovy, doesn't it?

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+@RestController
+@RequestMapping("/v1/config/deployments")
+public class DeploymentController {
+  @Autowired
+  HalconfigParser halyardConfig;
+
+  @RequestMapping(value = "/{deployment:.+}", method = RequestMethod.GET)
+  DeploymentConfiguration deploymentConfigurations(@PathVariable String deployment) throws UnrecognizedPropertyException {
+    Halconfig halconfig = halyardConfig.getConfig();
+    if (halconfig != null) {
```
Comment (ttomsu) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java]: Prefer early exits to deeply nested if blocks:

```
Halconfig halconfig = halyardConfig.getConfig();
if (halconfig == null) {
  return null
}
```

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      List<DeploymentConfiguration> matching = halconfig.getDeploymentConfigurations()
+          .stream()
+          .filter(d -> d.getName().equals(deployment))
+          .collect(Collectors.toList());
+
+      if (matching.size() == 0) {
+        throw new DeploymentNotFoundExecption(deployment);
+      } else if (matching.size() > 1) {
+        throw new DuplicateDeploymentException(deployment);
+      } else {
```
Comment (ttomsu) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java]: Drop `else` block

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    if (halconfig != null) {
+      List<DeploymentConfiguration> matching = halconfig.getDeploymentConfigurations()
+          .stream()
+          .filter(d -> d.getName().equals(deployment))
+          .collect(Collectors.toList());
+
+      if (matching.size() == 0) {
+        throw new DeploymentNotFoundExecption(deployment);
+      } else if (matching.size() > 1) {
+        throw new DuplicateDeploymentException(deployment);
```
Comment (ttomsu) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java]: Seems odd to me to throw an exception on a simple GET command, even if there are multiple matches. Are there other ways to enforce uniqueness of the `deployment` id?

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      }
+    } else {
+      return null;
+    }
+  }
+
+  @RequestMapping(value = "/", method = RequestMethod.GET)
+  List<DeploymentConfiguration> deploymentConfigurations() throws UnrecognizedPropertyException {
+    Halconfig halconfig = halyardConfig.getConfig();
+    if (halconfig != null) {
```
Comment (ttomsu) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java]: Simplify like above

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      return halconfig.getDeploymentConfigurations();
+    } else {
+      return null;
+    }
+  }
+
+  @Data
```
Comment (ttomsu) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java]: These aren't @Data classes, and probably don't need the equals and hash code either. You probably just want to make deployment name `private final` and use `@RequiredArgsConstructor`

Comment (lwander) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/DeploymentController.java]: `deploymentName` needs to be accessible for a user to make sense of what's wrong, so they are `@Data`. `EqualsAndHashCode` comes from the fact that I don't use the super, so it shouldn't be taken into account when performing equality checks.

**Pull request merged**